First pass of garbage collection command

aronasorman opened the pull request:
Create a new function called `clean_up_contentnodes`. This does the following:

- delete all ContentNode objects associated with the orphan tree.
- delete all File records associated with those ContentNodes.
- delete all file objects on storage associated with those Files.

We also create a minimal management command, called `garbage_collect`, that
simply calls `clean_up_contentnodes` for now.

Note that this is just the app side of deletion. Next PR will involve running this in a cron job in production.

jayoshih commented on contentcuration/contentcuration/tests/test_garbage_collect.py:
```diff
@@ -0,0 +1,101 @@
+from base import StudioTestCase
+
+
+THREE_MONTHS_AGO = datetime.now() - timedelta(days=93)
```
Do we want this value to be something under settings.py? (`days=settings.BUFFER_DELETE`)

jayoshih commented on contentcuration/contentcuration/utils/garbage_collect.py:
```diff
@@ -0,0 +1,63 @@
+from contentcuration.models import ContentNode, File
+
+
+TWO_WEEKS_AGO = datetime.now() - timedelta(days=14)
```
Same as comment above

jayoshih commented on contentcuration/contentcuration/utils/garbage_collect.py:
```diff
@@ -0,0 +1,63 @@
+
+    """
+    nodes_to_clean_up = ContentNode.objects.filter(
+        parent_id=settings.ORPHANAGE_ROOT_ID,
```
Would it be better to get all the nodes at once (in particular, thinking of ricecooker trees that might have bigger subtrees). Maybe something like 

```
garbage_node = ContentNode.objects.get(pk=settings.ORPHANAGE_ROOT_ID)
nodes_to_clean_up = ContentNode.objects.exclude(parent=None) # exclude parent node
                   .filter(tree_id=garbage_node.tree_id, modified__gt=delete_older_than) # Note: did you mean gt?

jayoshih commented on contentcuration/contentcuration/utils/garbage_collect.py:
```diff
@@ -0,0 +1,63 @@
+        # string value from inside that singleton tuple
+        file_path = f[0]
+        # call the storage's delete method on each file, one by one
+        storage.delete(file_path)
```
Might want to be careful if user is on develop.studio and deletes a file- will also delete on master. Probably okay to leave out for now

jayoshih replied:
Separate out buckets

aronasorman commented on contentcuration/contentcuration/tests/test_garbage_collect.py:
```diff
@@ -0,0 +1,101 @@
+        # orphan tree, should still exist after cleanup
+        legit_tree = ContentNode.objects.create(
+            kind_id="Topic",
+        )
```
Do you mean, we need to make sure that files associated with non-stale contentnodes don't get deleted? I believe that's what we do here.

kollivier replied:
Ah, sorry, I read legit and thought "a legit node in Studio, i.e. with a parent" but clearly the code doesn't do that. So I'll reframe my request as "add a ContentNode in a different tree with a date of THREE_MONTHS_AGO to ensure isn't deleted with the orphanage nodes". :) Basically, I think it's important to test that the garbage collector doesn't actually delete anything outside of its own tree, as really bad things could happen if we got a regression like that.

aronasorman replied:
@kollivier added a new test that makes an old, but legitimate tree, and checks that it's not deleted: https://github.com/learningequality/studio/pull/856/files#diff-acfc219a7c8f39b84d5f57bd92268c2dR105

aronasorman commented:
Addressed the comments related to using `modified` instead of `created`, as well as moving the deletion threshold to `settings.py`. However, I don't think we should merge this until we've implemented a separation between production's and develop's buckets.

aronasorman commented:
Made a new test that creates an old, but legit tree and makes sure we don’t
delete that:
https://github.com/learningequality/studio/pull/856/files#diff-acfc219a7c8f39b84d5f57bd92268c2dR105



Aron

On August 3, 2018 at 1:31:49 PM, Kevin Ollivier (notifications@github.com)
wrote:

*@kollivier* commented on this pull request.
------------------------------

In contentcuration/contentcuration/tests/test_garbage_collect.py
<https://github.com/learningequality/studio/pull/856#discussion_r207659729>:

> +        requests.head(file_url).raise_for_status()
+        clean_up_contentnodes()
+
+        # there should be no file object in the DB
+        assert File.objects.count() == 0
+
+    def test_doesnt_delete_nonorphan_files_and_contentnodes(self):
+        """
+        Make sure that clean_up_contentnodes doesn't touch non-orphan files and
+        contentnodes. Bad things will happen if we do.
+        """
+        # this legit tree, since it's not attached to our
+        # orphan tree, should still exist after cleanup
+        legit_tree = ContentNode.objects.create(
+            kind_id="Topic",
+        )

Ah, sorry, I read legit and thought "a legit node in Studio, i.e. with a
parent" but clearly the code doesn't do that. So I'll reframe my request as
"add a ContentNode in a different tree with a date of THREE_MONTHS_AGO to
ensure isn't deleted with the orphanage nodes". :) Basically, I think it's
important to test that the garbage collector doesn't actually delete
anything outside of its own tree, as really bad things could happen if we
got a regression like that.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/learningequality/studio/pull/856#discussion_r207659729>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AALt00M-tisl0VbMtb0iJRUnobgQrDAdks5uNLM1gaJpZM4VtNRa>
.

aronasorman commented:
Merged #861, a prerequisite for this.

I uploaded a new [PDF](develop.studio.learningequality.org/content/storage/a/d/ad9fd2f7b94fa8d5590488c26580b68e.pdf) on develop.studio. I changed the URL from `develop.studio` to just `studio`, and it gave a 404. Which means our buckets are now separate!

Merging this in once the tests pass.
